### PR TITLE
PlayState: Add Windows notifications

### PR DESCRIPTION
**Pull request by BanCrash**
I have verified that:
- [x] These changes work, by building the extension and testing.
- [x] That the changes comply with the [rules](https://github.com/darklinkpower/PlayniteExtensionsCollection#contributing) indicated in the repository.
- [x] Pull request is targeting `master` branch.

Changes:
- Added a setting for set a hotkey to show information about the current session of the game.
- Added a setting for use Windows notifications instead Playnite overlay system.
- Show information about play time on the notifications when Windows notifications are enabled.

As stated on the new settings note, in order to Windows notifications be showed it requires that Playnite is added to the priority list if the Focuss Assist is enabled when playing a game.

The notifications will be still showed after finishing the game since currentGame doesn't get deleted after finishing the game. I don't see any reason to not remove it on OnGameStopped event. Could you confirm that this is the case so I'll delete currentGame variable when finishing the game in order to avoid notifications about an ended game?

### CODE REVIEW

File: source/Generic/PlayState/PlayState.cs [hunk 1]
```diff
@@ -242,6 +267,14 @@ public override void OnGameStarted(OnGameStartedEventArgs args)
                 return;
             }
 
+            playedDateTime = DateTime.Now;
+
+            showWindowsNotificationsStyle = false;
+            if (settings.Settings.GlobalShowWindowsNotificationsStyle)
+            {
+                showWindowsNotificationsStyle = true;
+            }
```
[the remark this answers is not in the document]
Comment (darklinkpower) [reply source/Generic/PlayState/PlayState.cs:276]: A class could be created that has the info used by the extension, maybe like this

```csharp
    class PlaystateData
    {
        public id DatabaseId { get; set; }
        public datetime StartDate { get; set; }
        public Stopwatch Stopwatch { get; set; }
    }
```

This could replace the `List<Tuple<Guid, Stopwatch>>` list and use a `List<PlaystateData>` list instead to use in the extension. Maybe even put the revelant methods inside the class, like the ones to get the current playtime or so, but that's just an idea.

File: source/Generic/PlayState/PlayState.cs [hunk 2]
```diff
@@ -392,6 +425,175 @@ private void WindowClosed(object sender, EventArgs e)
...
+        /// - "suspended" for suspend process and playtime<br/>
+        /// - "playtimeSuspended" for suspend playtime<br/>
+        /// - "information" for showing the actual status<br/>
+        /// </param>
+        /// </summary>
+        private void ShowNotification(string status)
+        {
+            switch (status)
+            {
+                case "resumed": // for resuming process and playtime
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.cs:499]: Change to enums please

File: source/Generic/PlayState/PlayState.cs [hunk 3]
```diff
@@ -392,6 +425,175 @@ private void WindowClosed(object sender, EventArgs e)
...
+                    }
+                    else
+                    {
+                        // Windows notification only admits 3 .Addtext properties, so for more text lines you have to use {Environment.NewLine}
+                        new ToastContentBuilder()
+                            .AddText(currentGame.Name) // First AddText field will act as a title
+                            .AddText(ResourceProvider.GetString("LOCPlayState_StatusResumedMessage"))
+                            .AddText($"{ResourceProvider.GetString("LOCPlayState_Playtime")} {GetHoursString(GetRealPlaytime())}" +
+                                $"{Environment.NewLine}{ResourceProvider.GetString("LOCPlayState_TotalPlaytime")} {GetHoursString(GetRealPlaytime() + currentGame.Playtime)}")
+                            .AddHeroImage(new Uri(Path.Combine(PlayniteApi.Paths.ConfigurationPath, "library/files", currentGame.BackgroundImage))) // Show game image in the notification
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.cs:512]: Don't combine the path here. Use this method instead https://playnite.link/docs/master/api/Playnite.SDK.IGameDatabaseAPI.html#Playnite_SDK_IGameDatabaseAPI_GetFullFilePath_System_String_

Example: https://playnite.link/docs/master/tutorials/extensions/library.html#exporting-game-cover

Also, are the toasts compatible with urls? In case it doesn't, then that needs to be handled in case the game has an url set for the image

File: source/Generic/PlayState/PlayState.cs [hunk 4]
```diff
@@ -429,11 +631,11 @@ private void SwitchGameState()
                         isSuspended = false;
                         if (processesSuspended)
                         {
-                            ShowSplashWindow(ResourceProvider.GetString("LOCPlayState_StatusResumedMessage"));
+                            ShowNotification("resumed");
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.cs:634]: Enums

File: source/Generic/PlayState/PlayState.csproj [hunk 5]
```diff
@@ -12,6 +12,8 @@
     <TargetFrameworkVersion>v4.6.2</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
     <Deterministic>true</Deterministic>
+    <NuGetPackageImportStamp>
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.csproj:15]: What is this?

File: source/Generic/PlayState/PlayState.csproj [hunk 6]
```diff
@@ -31,9 +33,6 @@
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Playnite.SDK, Version=6.2.2.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>packages\PlayniteSDK.6.2.2\lib\net462\Playnite.SDK.dll</HintPath>
-    </Reference>
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.csproj:36]: Why was this removed?

File: source/Generic/PlayState/PlayState.csproj [hunk 7]
```diff
@@ -68,7 +67,6 @@
     <None Include="extension.yaml">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
-    <None Include="packages.config" />
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.csproj:71]: Why was this removed?

File: source/Generic/PlayState/PlayState.csproj [hunk 8]
```diff
@@ -93,6 +91,17 @@
...
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Toolkit.Uwp.Notifications">
+      <Version>7.1.2</Version>
+    </PackageReference>
+    <PackageReference Include="PlayniteSDK">
+      <Version>6.2.2</Version>
+    </PackageReference>
+    <PackageReference Include="System.Management">
+      <Version>5.0.0</Version>
+    </PackageReference>
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.csproj:103]: Is this reference used in the project?

File: source/Generic/PlayState/PlayStateSettings.cs [hunk 9]
```diff
@@ -56,12 +64,26 @@ public PlayStateSettingsViewModel(PlayState plugin)
...
+                    
+                }
+                if (Settings.SavedInformationHotkeyGesture.Modifiers == ModifierKeys.None)
+                {
+                    // Due to a bug in previous version that allowed 
+                    // to save gestures without modifiers, this
+                    // should be done to restore the default ModifierKeys
+                    Settings.SavedInformationHotkeyGesture.Modifiers = (ModifierKeys)5;
+                    modifiedSettings = true;
+                }
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayStateSettings.cs:84]: Why is this being done for the new hotkeys? The bug should not exist there right?

File: source/Generic/PlayState/packages.config [hunk 10]
```diff
@@ -1,5 +0,0 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<packages>
-  <package id="PlayniteSDK" version="6.2.2" targetFramework="net462" />
-  <package id="System.Management" version="5.0.0" targetFramework="net462" />
-</packages>
```
Comment (darklinkpower) [inline source/Generic/PlayState/packages.config:5]: Why was this removed?

**Review by darklinkpower — CHANGES_REQUESTED** [review]

File: source/Generic/PlayState/Localization/en_US.xaml [hunk 11]
```diff
@@ -17,17 +17,29 @@ Games will be suspended when using the configured hotkey.</sys:String>
...
     <sys:String x:Key="LOCPlayState_SettingSubstractSuspendedPlaytimeOnStoppedTooltip">Helps to keep a more accurate playtime value in games by only keeping the time the game was not suspended count towards the game playtime</sys:String>
     <sys:String x:Key="LOCPlayState_SettingSubstractOnlyNonLibraryGamesLabel">Only execute for games that are not managed by library plugins</sys:String>
     <sys:String x:Key="LOCPlayState_SettingGlobalOnlySuspendPlaytimeLabel">Only suspend playtime count globally and not the game processes</sys:String>
     <sys:String x:Key="LOCPlayState_SettingGlobalOnlySuspendPlaytimeTooltip" xml:space="preserve">By enabling this option, all games won't be suspended when using the configured hotkey and only the time will be counted to substracted from the game playtime when it stops.
     
 This can also be applied individually per game by adding a feature named "[PlayState] Suspend playtime only" to a game</sys:String>
+    <sys:String x:Key="LOCPlayState_SettingGlobalShowWindowsNotificationsStyleLabel">Show Windows notifications instead Playnite overlay</sys:String>
```
[the remark this answers is not in the document]
Comment (darklinkpower) [reply source/Generic/PlayState/Localization/en_US.xaml:35]: I also think it should be changed "Playnite overlay" -> "Playnite notification window" since it's not an overlay. If you like and agree, perhaps changing it to 

> Show native Windows notifications instead of Playnite notification window

[the remark this answers is not in the document]
Comment (BanCrash) [reply source/Generic/PlayState/Localization/en_US.xaml:35]: I like your change yeah, I will replace it.

File: source/Generic/PlayState/PlayState.cs [hunk 12]
```diff
@@ -100,21 +104,38 @@ internal bool RegisterGlobalHotkey()
...
                     PlayniteApi.Notifications.Add(new NotificationMessage(Guid.NewGuid().ToString(),
                         "PlayState: " + string.Format(ResourceProvider.GetString("LOCPlayState_NotificationMessageHotkeyRegisterFailed"), settings.Settings.SavedHotkeyGesture),
                         NotificationType.Error));
-                    logger.Error($"Failed to register configured Hotkey {settings.Settings.SavedHotkeyGesture}.");
+                    logger.Error($"Failed to register configured pause/resume Hotkey {settings.Settings.SavedHotkeyGesture}.");
                 }
 
-                return registered;
+                // Information Hotkey
+                var registered2 = HotkeyHelper.RegisterHotKey(handle, HOTKEY_ID, settings.Settings.SavedInformationHotkeyGesture.Modifiers.ToVK(), (uint)KeyInterop.VirtualKeyFromKey(settings.Settings.SavedInformationHotkeyGesture.Key));
```
[the remark this answers is not in the document]
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs:124]: The new hotkey is for just showing information about the actual state, not changes the state of paused/resumed. 

About the second, I didn't understand well the logic behind the hotkey system tbh, but in my tests it's working as expected.

File: source/Generic/PlayState/PlayState.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs:276]: I just followed the code as paused/resumed setting was being handled there too. But you are right, should be on the ShowNotification method.

I will see if I can make work the second comment.

File: source/Generic/PlayState/PlayState.cs (hunk 3 above)
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs:512]: I checked but didn't find a better way, thanks for pointing me out.

The toasts are compatible with links, in their doc they have this example:
`.AddInlineImage(new Uri("https://picsum.photos/360/202?image=883"))`

https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/send-local-toast?tabs=uwp

File: source/Generic/PlayState/PlayState.csproj (hunk 5 above)
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.csproj:15]: Aarrg, a file that showed the migration from packages.config to ReferencePackage. I removed the file but forgot about the reference here.

File: source/Generic/PlayState/PlayState.csproj (hunk 6 above)
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.csproj:36]: Was automatically removed by Visual Studio when changing to package reference: https://docs.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference

This change was necessary because it didn't compile after adding the notifications dependencies without doing so.

File: source/Generic/PlayState/PlayState.csproj (hunk 7 above)
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.csproj:71]: Was automatically removed by Visual Studio when changing to package reference: https://docs.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference

This change was necessary because it didn't compile after adding the notifications dependencies without doing so.

File: source/Generic/PlayState/PlayState.csproj (hunk 8 above)
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.csproj:103]: It was added automatically when adding the dependencies for Windows notifications, so I assume yes. Anyway I will make a test without it just to make sure.

File: source/Generic/PlayState/PlayStateSettings.cs (hunk 9 above)
Comment (BanCrash) [reply source/Generic/PlayState/PlayStateSettings.cs:84]: Like I said before I don't understand well the logic behind hotkeys. I thought this was executed after editing hotkeys all the time. I will remove it then.

File: source/Generic/PlayState/packages.config (hunk 10 above)
Comment (BanCrash) [reply source/Generic/PlayState/packages.config:5]: Was automatically removed by Visual Studio when changing to package reference: https://docs.microsoft.com/en-us/nuget/consume-packages/migrate-packages-config-to-package-reference

This change was necessary because it didn't compile after adding the notifications dependencies without doing so.

Comment (BanCrash) [reply source/Generic/PlayState/packages.config]: Okay, seems that it wasn't necessary. I reverted this change and added notifications on packages.config file.

File: source/Generic/PlayState/PlayState.cs [hunk 13]
```diff
@@ -134,6 +155,10 @@ private IntPtr GlobalHotkeyCallback(IntPtr hWnd, int msg, IntPtr wParam, IntPtr
                         {
                             SwitchGameState();
                         }
+                        else if (vkey == (uint)KeyInterop.VirtualKeyFromKey(settings.Settings.SavedInformationHotkeyGesture.Key))
+                        {
+                            ShowNotification("information");
```
[the remark this answers is not in the document]
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs]: Done

File: source/Generic/PlayState/PlayState.cs (hunk 3 above)
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs]: Done

File: source/Generic/PlayState/PlayState.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs:125]: I checked again. Like I've said I don't know the logic behind, but in my tests this is working as expected. If I use the hotkey for Pause/Resume, it does that, and if I use the hotkey for information, it doesn't resume but only show the information.

File: source/Generic/PlayState/PlayState.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs]: Done

**Comment by BanCrash** [comment]
I think I fixed all the issues @darklinkpower 
I create the new class as you suggested, and now when currentGame is finished, it will set currentGame to the last executed game if more than one were executing at the same time.

**Comment by darklinkpower** [comment]
I wanted to build it before giving the last check but the Microsoft.Toolkit namespace was not being detected. Upon uninstalling the package and installing again, I get this

image

---

Another thing, here it says it's only for Windows 10. Can you add a check and only allow/try to send toasts if the user is on Windows 10? Could be a value set in the constructor

image

File: source/Generic/PlayState/PlayState.cs [hunk 14]
```diff
@@ -494,6 +488,12 @@ private string GetHoursString(ulong playtimeSeconds)
         /// </summary>
         private void ShowNotification(string status)
         {
+            showWindowsNotificationsStyle = false;
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.cs:489]: can be simplified as

```csharp
showWindowsNotificationsStyle = settings.Settings.GlobalShowWindowsNotificationsStyle
```

Also, is it needed to store the value in the plugin itself? Or can it be a variable inside this method like

```csharp
var showWindowsNotificationsStyle = settings.Settings.GlobalShowWindowsNotificationsStyle
```

File: source/Generic/PlayState/PlayState.cs [hunk 15]
```diff
@@ -771,29 +773,32 @@ public override void OnGameStopped(OnGameStoppedEventArgs args)
...
+            if (playStateData.Any(x => x.Game.Id == game.Id && x.Stopwatch != null))
             {
-                logger.Debug($"A stopwatch for game {game.Name} with id {game.Id} already exists");
+                logger.Debug($"Data for game {game.Name} with id {game.Id} already exists");
             }
             else
             {
-                stopwatchList.Add(new Tuple<Guid, Stopwatch>(game.Id, new Stopwatch()));
-                logger.Debug($"Stopwatch for game {game.Name} with id {game.Id} was created");
+                playStateData.Add(new PlayStateData(game, DateTime.Now, new Stopwatch()));
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.cs:787]: Do you think it would make sense to get the `DateTime.Now` value and create the stopwatch inside the class constructor? I think it would make more sense since this is something that doesn't need outside reference and will simplify the object creation

File: source/Generic/PlayState/PlayState.cs [hunk 16]
```diff
@@ -771,29 +773,32 @@ public override void OnGameStopped(OnGameStoppedEventArgs args)
...
+                logger.Debug($"Data for game {game.Name} with id {game.Id} was created");
             }
         }
 
-        private void RemoveGameStopwatchTuple(Game game)
+        private void RemoveGame(Game game)
         {
-            var tuple = stopwatchList.FirstOrDefault(x => x.Item1 == game.Id);
-            if (tuple != null)
+            if (playStateData.Any(x => x.Game.Id == game.Id))
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayState.cs:794]: Could use

```csharp
var gameData = FirstOrDefault(x => x.Game.Id == game.Id);
if (gameData != null)
{
    playStateData.Remove(gameData);
```

To prevent iterating twice

**Review by darklinkpower — CHANGES_REQUESTED** [review]

**Comment by BanCrash** [comment]
> I wanted to build it before giving the last check but the Microsoft.Toolkit namespace was not being detected. Upon uninstalling the package and installing again, I get this
> 
> image
> 
> Another thing, here it says it's only for Windows 10. Can you add a check and only allow/try to send toasts if the user is on Windows 10? Could be a value set in the constructor
> 
> image

Yes, that's the issue that force me to do the first change about removing packages.config and so. But the funny thing is now it's working for me. I will do the switch again then.

Yes, you are right, only wor W10. I will check it.

File: source/Generic/PlayState/PlayState.cs [hunk 17]
```diff
@@ -392,16 +417,196 @@ private void WindowClosed(object sender, EventArgs e)
...
+        /// </summary>
+        private ulong GetRealPlaytime()
+        {
+            var suspendedTime = playStateData.FirstOrDefault(x => x.Game.Id == currentGame.Id)?.Stopwatch.Elapsed;
+            ulong elapsedSeconds = 0;
+            if (suspendedTime != null)
+            {
+                elapsedSeconds = Convert.ToUInt64(suspendedTime.Value.TotalSeconds);
+            }
+            return Convert.ToUInt64(DateTime.Now.Subtract(playStateData.FirstOrDefault(x => x.Game.Id == currentGame.Id).StartDate).TotalSeconds) - elapsedSeconds;
```
[the remark this answers is not in the document]
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs:431]: Shouldn't be the case since the comprobations should be made before, so yeah, I will go with First instead FirstOrDefault

File: source/Generic/PlayState/PlayState.cs (hunk 14 above)
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs:489]: I did that way for following the previous code, but the second case you said I think it's better. I will use it.

File: source/Generic/PlayState/PlayState.cs (hunk 15 above)
Comment (BanCrash) [reply source/Generic/PlayState/PlayState.cs:787]: Shouldn't be any issue doing so, I will test it.

**Comment by BanCrash** [comment]
@darklinkpower all changes done. I restore the switch to PackageReference, should compile for you now. 
For W10 or W11 comprobation I had to check registry since for some reason Environment.OSVersion is not working.

File: source/Generic/PlayState/PlayStateSettingsView.xaml.cs [hunk 18]
```diff
@@ -149,5 +154,11 @@ private void SetInformationHotkey(KeyEventArgs e, TextBox tbInformationHotkey, B
             setInformationHotkeyButton.Content = ResourceProvider.GetString("LOCPlayState_SettingChangeHotkeyButtonLabel");
         }
 
+        private bool IsWindows10Or11()
+        {
+            var productName = (string) Microsoft.Win32.Registry.LocalMachine.OpenSubKey(@"Software\Microsoft\Windows NT\CurrentVersion")?.GetValue("productName");
```
Comment (darklinkpower) [inline source/Generic/PlayState/PlayStateSettingsView.xaml.cs:159]: The key needs to be disposed by using Dispose() method or the `using (...` similar to here https://github.com/darklinkpower/PlayniteExtensionsCollection/blob/24798e311c1db8dfd8eebbd5d1240ad842b20a28/source/Generic/SpecialKHelper/SpecialKHelper.cs#L194-L212 but I will take care of it after merging

**Comment by darklinkpower** [comment]
This should also go here since it's about that specific option but I'll take care of this 
image

**Comment by darklinkpower** [comment]
And there should be margin here. I'll take care of it 
image

**Comment by darklinkpower** [comment]
The message should be different when the action was pause/unpause game or is just displaying information since currently it's not clear if something was done

image

**Comment by darklinkpower** [comment]
I'll merge the PR, I'll make some modifications to some things I think need small changes and release afterwards. Thanks for contributing!

**Comment by BanCrash** [comment]
> I'll merge the PR, I'll make some modifications to some things I think need small changes and release afterwards. Thanks for contributing!

Sure, no problem, thanks!

File: source/Generic/PlayState/PlayStateSettingsView.xaml.cs [hunk 19]
```diff
@@ -25,6 +26,10 @@ public PlayStateSettingsView()
             InitializeComponent();
             TbHotkey.PreviewKeyDown += HotkeyTextBox_PreviewKeyDown;
             TbInformationHotkey.PreviewKeyDown += InformationHotkeyTextBox_PreviewKeyDown;
+            if (!IsWindows10Or11())
```
[the remark this answers is not in the document]
Comment (darklinkpower) [reply source/Generic/PlayState/PlayStateSettingsView.xaml.cs:29]: Got it. I changed it in https://github.com/darklinkpower/PlayniteExtensionsCollection/commit/05da1f1d947b1dcba5ae5b08461928dcdb2b8abb to make it usable during the evaluation to select the notification style

And in here I disabled the option in settings with MVVM pattern which is preferable https://github.com/darklinkpower/PlayniteExtensionsCollection/commit/804a886d5cfb62897e05fe4f9082e6c08822b432

[the remark this answers is not in the document]
Comment (BanCrash) [reply source/Generic/PlayState/PlayStateSettingsView.xaml.cs:29]: Sure, perfect then!
